viewer/src/datamodels/cad/sector/CachedRepository.ts

```diff
@@ -100,100 +98,95 @@ export class CachedRepository implements Repository {
   }
 
   getLoadingStateObserver(): Observable<boolean> {
-    return this._isLoadingSubject.pipe(distinctUntilChanged(), share());
+    return this._loadingCounter.observeCount().pipe(
+      distinctUntilChanged(),
+      map(count => count != 0)
+    );
   }
 
   // TODO j-bjorne 16-04-2020: Should look into ways of not sending in discarded sectors,
   // unless we want them to eventually set their priority to lower in the cache.
 
-  loadSector(): OperatorFunction<WantedSector[], ConsumedSector> {
+  loadSector(): OperatorFunction<WantedSector, ConsumedSector> {
     return pipe(
-      subscribeOn(asapScheduler),
-      switchMap(wantedSectorsArray => {
-        return from(wantedSectorsArray).pipe(
-          tap(_ => {
-            this._isLoadingSubject.next(true);
-          }),
-          publish(wantedSectorsObservable => {
-            const simpleAndDetailedObservable = wantedSectorsObservable.pipe(
-              filter(
-                wantedSector =>
-                  wantedSector.levelOfDetail === LevelOfDetail.Simple ||
-                  wantedSector.levelOfDetail === LevelOfDetail.Detailed
-              ),
-              this.loadSimpleAndDetailedSector()
-            );
-
-            const discardedSectorObservable = wantedSectorsObservable.pipe(
-              filter(wantedSector => wantedSector.levelOfDetail === LevelOfDetail.Discarded),
-              map(wantedSector => ({ ...wantedSector, group: undefined } as ConsumedSector))
-            );
-
-            return merge(simpleAndDetailedObservable, discardedSectorObservable);
-          }),
-          finalize(() => this._isLoadingSubject.next(false))
-        );
-      }),
-      finalize(() => {
-        this.clearCaches();
-      })
+      this._loadingCounter.add(),
+      flatMap(sector => {
+        if (this._consumedSectorCache.has(this.wantedSectorCacheKey(sector))) {
+          return this._consumedSectorCache.get(this.wantedSectorCacheKey(sector));
+        } else if (sector.levelOfDetail == LevelOfDetail.Detailed) {
+          return this.loadDetailedSectorFromNetwork(sector);
+        } else if (sector.levelOfDetail == LevelOfDetail.Simple) {
+          return this.loadSimpleSectorFromNetwork(sector);
+        } else if (sector.levelOfDetail == LevelOfDetail.Discarded) {
+          return of(sector).pipe(map(wantedSector => ({ ...wantedSector, group: undefined } as ConsumedSector)));
+        } else {
+          throw new Error('Unhandled case');
+        }
+      }, this._concurrentNetworkOperations),
+      this._loadingCounter.remove(),
```

Comment:
What happens if the `flatMap`-thing throws? Are we forever in loading-state because the counter is never decremented?

Reply:
Finalize runs if the observable completes, and that should reset the counter to 0 and loading. The reason why I introduced this is due to loading not being very helpful and I want to change it to Loading X sectors or something, so that there is a bit more indication that something is happening :)